chain/network/src/test_loop.rs

```diff
@@ -1,4 +1,301 @@
+            None
+        }
+        _ => Some(request),
+    })
+}
+
+fn network_message_to_partial_witness_handler(
+    my_account_id: &AccountId,
+    partial_witness_senders: HashMap<AccountId, PartialWitnessSenderForNetwork>,
+) -> NetworkRequestHandler {
+    let my_account_id = my_account_id.clone();
+    Arc::new(move |request| match request {
+        NetworkRequests::ChunkStateWitnessAck(target, witness_ack) => {
+            // TODO(testloop): Convert account check to a panic.
+            if target != my_account_id {
+                let sender = partial_witness_senders.get(&target).unwrap();
+                sender.send(ChunkStateWitnessAckMessage(witness_ack));
+            }
+            None
+        }
+
+        NetworkRequests::PartialEncodedStateWitness(validator_witness_tuple) => {
+            for (target, partial_witness) in validator_witness_tuple.into_iter() {
+                // TODO(testloop): Convert account check to a panic.
+                if target != my_account_id {
+                    let sender = partial_witness_senders.get(&target).unwrap();
+                    sender.send(PartialEncodedStateWitnessMessage(partial_witness));
+                }
+            }
+            None
+        }
+        NetworkRequests::PartialEncodedStateWitnessForward(chunk_validators, partial_witness) => {
+            for target in chunk_validators {
+                // TODO(testloop): Convert account check to a panic.
+                if target != my_account_id {
+                    let sender = partial_witness_senders.get(&target).unwrap();
+                    sender.send(PartialEncodedStateWitnessForwardMessage(partial_witness.clone()));
+                }
+            }
+            None
+        }
+        _ => Some(request),
+    })
+}
+
+fn network_message_to_state_snapshot_handler() -> NetworkRequestHandler {
+    Arc::new(move |request| match request {
+        NetworkRequests::SnapshotHostInfo { .. } => None,
+        _ => Some(request),
+    })
+}
+
+#[derive(Default, Clone)]
+struct RouteLookup(Arc<Mutex<HashMap<CryptoHash, AccountId>>>);
+
+impl RouteLookup {
+    fn new() -> Self {
+        Self(Arc::new(Mutex::new(HashMap::new())))
+    }
+
+    fn add_route(&self, return_account_id: &AccountId) -> CryptoHash {
+        let mut guard = self.0.lock().unwrap();
+        let route_id = CryptoHash::hash_borsh(guard.len());
+        guard.insert(route_id, return_account_id.clone());
+        route_id
+    }
+
+    fn get_destination(&self, route_id: CryptoHash) -> AccountId {
+        let guard = self.0.lock().unwrap();
+        guard.get(&route_id).unwrap().clone()
+    }
+}
+
+fn network_message_to_shards_manager_handler(
+    clock: Clock,
+    my_account_id: &AccountId,
+    shards_manager_senders: HashMap<AccountId, Sender<ShardsManagerRequestFromNetwork>>,
+) -> Arc<dyn Fn(NetworkRequests) -> Option<NetworkRequests>> {
+    static ROUTE_LOOKUP: Lazy<RouteLookup> = Lazy::new(RouteLookup::new);
```

Comment:
There seems to be an issue: if we run many tests at once, `ROUTE_LOOKUP` will be initialised only once and may cause chaos in network routes related to different TestLoop instances.
Instead, can we create RouteLookup in `TestLoopBuilder` and pass it to all `TestLoopPeerManagerActor`s in `setup_network`?

Reply:
Hmm, yes, that's what I was originally doing but it didn't look too elegant. I thought it might be fine to have a static variable here as (1) it's behind a lock, and (2) we are only adding new entries for the route, not removing anything in a way that can cause conflict. Ideally multiple instances should be able to use the same route lookup.

What is route lookup? It's basically the mapping from who sent the `PartialEncodedChunkRequest` to whom and who to respond back to with the `PartialEncodedChunkResponse`. I'll add this to the documentation.

I'm fine with implementing any approach but mildly prefer the current one (unless proven to not work). Let me know your thoughts?

Reply:
Sure, let's leave it as static, but also please add comment why it is fine to reuse across the same binary (e.g. same route hash means... same target account?)

Reply:
Added comments to explain this